Update rasterio

klinkow opened the pull request:
Updates dependency Rasterio from version `>=1.0a12` to `^1.4.3`. Updates project to use Poetry, replacing `Pipfile` with `pyproject.toml`.

klinkow commented on trefoil/netcdf/tests/test_crs.py:
```diff
@@ -34,7 +34,6 @@ def test_get_crs(tmpdir):
 
     out_data = CRS.from_string(out_proj4).to_dict()
 
-    assert len(out_data) == 8  # There should be 8 parameters
```
This assertion failed (6 != 8), but it's hard to know what 8 things were wanted. More direct assertions of the contents of `out_data` follow.

nikmolnar replied:
Weird, when I run this code in the interpreter, I get a dict with 10 items 🤔 Anyway, I think it's okay to remove this assertion, since in the next one, we're verifying the CRS from the dataset is what we expect.

klinkow replied:
Sorry, 10 is right. I remembered a difference of 2 when writing the comment and went the wrong direction.

klinkow commented on trefoil/netcdf/tests/test_crs.py:
```diff
@@ -56,7 +55,6 @@ def test_get_crs(tmpdir):
 
     out_data = CRS.from_string(out_proj4).to_dict()
 
-    assert len(out_data) == 4  # There should be 4 parameters
```
This assertion failed, but it's hard to know what 4 things were wanted. More direct assertions of the contents of `out_data` follow.

klinkow commented on trefoil/render/renderers/tests/test_renderers.py:
```diff
@@ -20,7 +20,6 @@ def test_stretched_renderer(tmpdir):
     assert renderer.name == 'stretched'
 
     img = renderer.render_image(data)
-    assert len(img.getpalette()) / 3 == 256
```
This assertion was failing, but I thought it wasn't needed if all of the subsequent assertions were passing.

nikmolnar replied:
How was it failing? The value didn't equal 256? Or it was causing an exception?

klinkow replied:
Thanks for the suggestion. The images look identical.

nikmolnar commented on pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+authors = ["Conservation Biology Institute"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
```
We need to add `six` to the dependencies. I tried importing a trefoil module and got an error about no module `six`.

klinkow replied:
Added

nikmolnar commented:
Have you tested this out with runs in SST? I'm getting blank result, and the raster produced seems to be all one value (128; probably the mask value since it shows completely invisible in the tool).

nikmolnar commented on pyproject.toml:
```diff
@@ -0,0 +1,32 @@
+[tool.poetry]
+name = "trefoil"
+version = "0.1.0"
```
This should be `0.4.0`.

klinkow commented:
> I'm getting blank result, and the raster produced seems to be all one value (128; probably the mask value since it shows completely invisible in the tool).

I'm not having any issues, but maybe I've left out a step in my testing. FYI since yesterday I've made the version update you suggested in this branch and also updated the ncdjango `test-trefoil-update` branch to target this branch instead of `main` (to make this testing easier). Here's my testing steps:

I added these changes to seedlot-selection-tool's `develop` branch (as of 7:50AM Pacific 6/24):
 
<img width="800" alt="image" src="https://github.com/user-attachments/assets/cbec5d2c-6cac-41c0-9a60-4ee829502583" />

I ran `poetry lock`, `poetry install` from seedlot-selection-tool, then `docker compose up`.

I ran the following configuration:

<img width="792" alt="image" src="https://github.com/user-attachments/assets/5cfcd5dc-6f5b-413a-bd2c-1d43eaf0ed3d" />
(also works adding a geographic constraint, but the results were small for a screenshot)

 Toggling on MAT layer:
<img width="755" alt="image" src="https://github.com/user-attachments/assets/e73d0e17-7f6b-4e22-9693-320e322211ac" />

klinkow commented:
Yep, I'm seeing the same errors I think. The task succeeds, but it can't serve the output NetCDF file as tiles. I get an error loading that file into QGIS.

Hmm... I'm thinking about how I can make a test that's better than logging out results along the way and comparing with working results. Maybe I'll start there and see where it takes me.

nikmolnar commented:
> Yep, I'm seeing the same errors I think. The task succeeds, but it can't serve the output NetCDF file as tiles. I get an error loading that file into QGIS.
> 
> Hmm... I'm thinking about how I can make a test that's better than logging out results along the way and comparing with working results. Maybe I'll start there and see where it takes me.

Sounds like a plan 👍

nikmolnar commented on pyproject.toml:
```diff
@@ -0,0 +1,32 @@
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = ">=3.10,<3.11"
```
Will this library really not work with Python 3.11+? It looks like rasterio supports 3.9+.

klinkow replied:
Nice catch. I updated this to `python = ">=3.9,<3.12"` after testing:
3.9 ✅  
3.10 ✅  
3.11 ✅ 
3.12 ❌ 
3.13 ❌

klinkow commented on trefoil/render/renderers/stretched.py:
```diff
@@ -112,7 +112,7 @@ def render_image(self, data, row_major_order=True):
 
         # derive palette index, and clip to [0, num_colors - 1]
-        stretched = ((values - self.min_value) * factor).astype(numpy.int)
+        stretched = ((values - self.min_value) * factor).astype(numpy.int_)
```
When running tests using `numpy.int`, I get:
```
FAILED trefoil/render/renderers/tests/test_renderers.py::test_stretched_renderer - AttributeError: module 'numpy' has no attribute 'int'.
```
According to [this SO thread](https://stackoverflow.com/questions/74946845/attributeerror-module-numpy-has-no-attribute-int): 

> numpy.int was deprecated in NumPy 1.20 and was removed in NumPy 1.24. 

I swapped in `int` as you suggested and it passes 🎉 Nice catch knowing the type of `numpy.int` (or `numpy.int_`).

nikmolnar commented on README.md:
````diff
@@ -42,16 +42,20 @@ We are using this on a variety of internal projects within the Conservation
-### No longer directly maintained / supported:
-On Windows, install the ones that require compiling from [Python Windows Packages](http://www.lfd.uci.edu/~gohlke/pythonlibs/).
-Then install the remainder using `pip`
+## Installation
+Trefoil is available on PyPi. To install, use:
 
+```text
+$ pip install trefoil
+```
````
Let's move "Installation" above the "Development" section since it's the more common thing people would want to do.
